internal/cache/db.go

```diff
@@ -0,0 +1,342 @@
+			case <-c.doClean:
+				func() {
+					c.mu.Lock()
+					defer c.mu.Unlock()
+
+					if err := c.db.Close(); err != nil {
+						slog.Warn(fmt.Sprintf("Could not close database %v", err))
+					}
+
+					db, err := openAndInitDB(dbPath, c.dirtyFlagPath)
+					if err != nil {
+						panic(fmt.Sprintf("CRITICAL: unrecoverable state: could not recreate database: %v", err))
+					}
+					c.db = db
+				}()
+			case <-c.quit:
+				return
+			}
+		}
+	}()
+	<-cleanupRoutineStarted
+
+	return &c, nil
+}
+
+// openAndInitDB open a pre-existing database and potentially intializes its buckets.
+// It cleans up any database previously marked as dirty or if it’s corrupted.
+func openAndInitDB(path, dirtyFlagPath string) (*bbolt.DB, error) {
+	if _, err := os.Stat(dirtyFlagPath); err == nil {
+		cleanDatabase(path, dirtyFlagPath)
+	}
+
+	db, err := bbolt.Open(path, 0600, nil)
+	if err != nil {
+		if errors.Is(err, bbolt.ErrInvalid) {
+			cleanDatabase(path, dirtyFlagPath)
+			return nil, shouldRetryDBError{err: err}
+		}
+		return nil, fmt.Errorf("can't open database file: %v", err)
+	}
+	// Fail if permissions are not 0600
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return nil, fmt.Errorf("can't stat database file: %v", err)
+	}
+	perm := fileInfo.Mode().Perm()
+	if perm != 0600 {
+		return nil, fmt.Errorf("wrong file permission for %s: %o", path, perm)
+	}
+
+	// Create buckets
+	err = db.Update(func(tx *bbolt.Tx) error {
+		var allBucketsNames []string
+		for _, bucket := range allBuckets {
+			allBucketsNames = append(allBucketsNames, string(bucket))
+			if _, err := tx.CreateBucketIfNotExists(bucket); err != nil {
+				return err
+			}
+		}
+
+		// Clean up any unknown buckets
+		var bucketNamesToDelete [][]byte
+		err = tx.ForEach(func(name []byte, b *bbolt.Bucket) error {
+			if slices.Contains(allBucketsNames, string(name)) {
+				return nil
+			}
+			bucketNamesToDelete = append(bucketNamesToDelete, name)
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+		for _, bucketName := range bucketNamesToDelete {
+			// We are in a RW transaction.
+			_ = tx.DeleteBucket(bucketName)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return db, nil
+}
+
+// Close closes the db and signal the monitoring goroutine to stop.
+func (c *Cache) Close() error {
+	close(c.quit)
+	<-c.cleanupQuitted
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.db.Close()
+}
+
+// requestCleanDatabase ask for the clean goroutine to cleanup the database.
+// If we already have a pending request, do not block on it.
+func (c *Cache) requestCleanDatabase() {
+	if err := os.WriteFile(c.dirtyFlagPath, nil, 0600); err != nil {
+		slog.Warn(fmt.Sprintf("Could not write dirty file flag to signal cleaning up the database: %v", err))
+	}
+	select {
+	case c.doClean <- struct{}{}:
+	case <-time.After(time.Millisecond): // Let the time for the cleanup goroutine for the initial start.
+	}
+}
+
+func cleanDatabase(dbPath, dirtyFlagPath string) {
```

Comment:
I feel like this should be `clearDatabase` rather than `clean`. I think `clean` gives the impression that we do some operation on a few entries (like removing expired users) rather than removing the file completely, as is the case here.

Reply:
good one!